### ISSUE TITLE
Reports false positive with union of Protocols and Dict[str, Any] | None

### ISSUE THREAD

**Issue by XieJiSS** [microsoft/pyright#7324]
I think this looks similar to https://github.com/microsoft/pyright/issues/6957 but they are not identical, because my bug does not require generics or `TypeVar` to trigger. 

**Describe the bug**

False positive (by means of error) is reported with union of Protocols when one of the Protocol accepts `Dict[str, Any]` and the real function also accepts `Dict[str, Any] | ...`.

So basically, I have two protocols, defining functions that accepts either a `Dict[str, Any]` or a `None`. These protocols are named as `DictProtocol` and `NoneProtocol`, respectively. Then, I created a type `F`, which is the union type of these two protocols:

```python3
type F = DictProtocol | NoneProtocol
```

At this point, I'd like to declare a function `test` that can accepts all parameters that can be passed to any function of type `F`. So I write like this:

```python3
def test(x: Dict[str, Any] | None):
    if type(x) is dict:
        print("is dict")
    else:
        print("is None")
    pass

func_f: F = test
func_f({"1": 1})  # error
func_f(None)      # error
```

Despite getting two errors from Pylance, I ran this file and it prints `is dict` and `is None`, which is correct.

I can't figure out why Pylance stops me from doing this, so I changed `test` by removing `| None`:

```python3
def test(x: Dict[str, Any]):
    if type(x) is dict:
        print("is dict")
    else:
        print("not allowed")

func_f: F = test
func_f({"1": 1})  # ok
func_f(None)      # gets error, which is correct
```

This time Pylance works fine, and running the script in python lead to `is dict`, `not allowed`.

After playing with this for a while, I finally came up with 4 similar different cases:

**Case 1.** `test` accepts `Dict[str, Any] | None`

<img width="450" alt="image" src="https://github.com/microsoft/pyright/assets/24671280/d2036a48-10aa-4d34-b4ce-7680324fe881">

**Case 2.** `test` accepts `Dict[str, int] | None`

(cont.d) However, if we change the signature of `test` to `def test(_: Dict[str, int] | None)` then everything works fine:

<img width="450" alt="image" src="https://github.com/microsoft/pyright/assets/24671280/4122c81f-4a87-4339-9a76-295493dde3e3">

**Case 3.** Making `type F` contains a `Protocol` which accepts `Dict[str, Any] | None`

(cont.d) My first guess is that Pylance cannot narrow down `F` when given `test(_: Dict[str, Any] | None) -> None`, so I tried adding a `BothProtocol`, but no luck:

<img width="450" alt="image" src="https://github.com/microsoft/pyright/assets/24671280/a2ffa9b8-82f4-4865-9288-4982c95d48f4">

Note that even though `BothProtocol` matches `test`'s signature exactly, the type of `func_f` is still displayed as F, not `test(_: Dict[str, Any] | None) -> None` or `BothProtocol`.

**Case 4.**

(cont.d) However, if we remove `| None` from `test`'s signature, this time `func_f` can be narrowed as `DictProtocol`, and false positives disappear:

<img width="450" alt="image" src="https://github.com/microsoft/pyright/assets/24671280/3a780eb9-fc6a-476e-bc14-b9f9e2a44205">

Removing `BothProtocol` from case 4 yields the same result.

---

**Code or Screenshots**

Case 1:

```python
from __future__ import annotations
from typing import Dict, Any, Protocol

class DictProtocol(Protocol):
    def __call__(self, x: Dict[str, Any], /) -> Any:
        ...

class NoneProtocol(Protocol):
    def __call__(self, x: None, /) -> Any:
        ...

type F = DictProtocol | NoneProtocol

def test(_: Dict[str, Any] | None):
    pass

# on hover, func_f shows: func_f: F (!)
func_f: F = test
func_f({"1": 1})  # should not error, because test accepts Dict[str, Any].
func_f(None)      # should not error, because test accepts None.

test({"1": 1})    # ok
```

Case 2:

```python
from __future__ import annotations
from typing import Dict, Any, Protocol

class DictProtocol(Protocol):
    def __call__(self, x: Dict[str, Any], /) -> Any:
        ...

class NoneProtocol(Protocol):
    def __call__(self, x: None, /) -> Any:
        ...

type F = DictProtocol | NoneProtocol

def test(_: Dict[str, int] | None):
    pass

# on hover, func_f shows: def func_f(_: Dict[str, int] | None) -> None
func_f: F = test
func_f({"1": 1})  # ok
func_f(None)      # ok

test({"1": 1})    # ok
```

Case 3:

```python
from __future__ import annotations
from typing import Dict, Any, Protocol

class DictProtocol(Protocol):
    def __call__(self, x: Dict[str, Any], /) -> Any:
        ...

class NoneProtocol(Protocol):
    def __call__(self, x: None, /) -> Any:
        ...

class BothProtocol(Protocol):
    def __call__(self, x: Dict[str, Any] | None, /) -> Any:
        ...

type F = DictProtocol | NoneProtocol | BothProtocol

def test(_: Dict[str, Any] | None):
    pass

# on hover, func_f shows: func_f: F (!)
func_f: F = test
func_f({"1": 1})  # should not error, because test accepts Dict[str, Any].
func_f(None)      # should not error, because test accepts None.

test({"1": 1})    # ok
```

Case 4:

```python
from __future__ import annotations
from typing import Dict, Any, Protocol

class DictProtocol(Protocol):
    def __call__(self, x: Dict[str, Any], /) -> Any:
        ...

class NoneProtocol(Protocol):
    def __call__(self, x: None, /) -> Any:
        ...

class BothProtocol(Protocol):
    def __call__(self, x: Dict[str, Any] | None, /) -> Any:
        ...

type F = DictProtocol | NoneProtocol | BothProtocol

def test(_: Dict[str, Any]):
    pass

# on hover, func_f shows: func_f: DictProtocol (!)
func_f: F = test
func_f({"1": 1})  # does not error, which is correct (!)
func_f(None)      # got error here, which is also correct (!)

test({"1": 1})    # ok
```

Removing `BothProtocol` from case 4 yields the same result.

**VS Code extension or command-line**

I'm using Pylance provided by VS Code Python extention, version `v2024.1.10531008 (pre-release)`.

<img width="511" alt="image" src="https://github.com/microsoft/pyright/assets/24671280/bd98c99a-6f1b-416b-a4dc-f161a973d623">

**Comment by erictraut** [comment]
Pyright is working as intended here, and I don't consider this a bug.

When pyright encounters an assignment to a variable with a declared type, it attempts to narrow the type based on the assigned value. The narrow-on-assignment rules are complex because they need to consider cases where the declared value and/or the assigned value include an `Any` or `Unknown` type within them. The rules for type narrowing on assignment are not formalized in the typing spec, so it's up to each type checker implementation to decide how to do this. However, there are basic principles that are generally followed by type checkers with regard to narrowing types with `Any`. The behaviors you're seeing above are related to these narrowing behaviors.
